modules/table/src/main/java/org/apache/ignite/internal/table/distributed/storage/InternalTableImpl.java

```diff
@@ -1470,4 +1478,62 @@ private RuntimeException wrapReplicationException(Throwable e) {
 
         return e0;
     }
+
+    @Override
+    public @Nullable PendingComparableValuesTracker<HybridTimestamp> getPartitionSafeTimeTracker(int partitionId) {
+        return safeTimeTrackerByPartitionId.get(partitionId);
+    }
+
+    @Override
+    public @Nullable PendingComparableValuesTracker<Long> getPartitionStorageIndexTracker(int partitionId) {
+        return storageIndexTrackerByPartitionId.get(partitionId);
+    }
+
+    /**
+     * Updates the partition safe time tracker, if there was a previous one, closes it.
+     *
+     * @param partitionId Partition ID.
+     * @param newSafeTimeTracker New partition safe time tracker.
+     */
+    public void updatePartitionSafeTimeTracker(int partitionId, PendingComparableValuesTracker<HybridTimestamp> newSafeTimeTracker) {
+        PendingComparableValuesTracker<HybridTimestamp> previousSafeTimeTracker;
+
+        synchronized (updatePartitionMapsMux) {
+            Int2ObjectMap<PendingComparableValuesTracker<HybridTimestamp>> newSafeTimeTrackerMap = new Int2ObjectOpenHashMap<>(partitions);
+
+            newSafeTimeTrackerMap.putAll(safeTimeTrackerByPartitionId);
+
+            previousSafeTimeTracker = newSafeTimeTrackerMap.put(partitionId, newSafeTimeTracker);
+
+            safeTimeTrackerByPartitionId = newSafeTimeTrackerMap;
+        }
+
+        if (previousSafeTimeTracker != null) {
+            previousSafeTimeTracker.close();
```

Comment:
Do we have a test that makes sure that replaced trackers are closed?

Reply:
Create `org.apache.ignite.internal.table.distributed.storage.InternalTableImplTest#testUpdatePartitionTrackers`